.github/workflows/continuous-integration.yml

```diff
@@ -0,0 +1,266 @@
+
+  stable-php:
+    name: Determine Stable PHP Version for Non-Unit Test Checks
+    runs-on: ubuntu-latest
+    needs: [config, supported-versions-matrix]
+    outputs:
+      version: ${{ steps.stable-php.outputs.version }}
+    env:
+      CONFIG_JSON: ${{ needs.config.outputs.json }}
+    steps:
+      - id: stable-php
+        run: |
+          # This grabs the default stable PHP version for non-unit-test checks from the laminas.dev API.
+          # If the .github/workflows/continuous-integration-config.json file is
+          # present, and defines a string "stable-php" element, that value will
+          # be used.
+          DEFAULT_VERSION=$(curl -s https://laminas.dev/api/ci/stable-checks-version | jq -r '.version')
+          VERSION=$(echo "${CONFIG_JSON}" | jq -r 'def isstring(x): if x|type=="string" then x else empty end ; isstring(.["stable-php"])')
+          VERSION=${VERSION:-$DEFAULT_VERSION}
+          echo "PHP version discovered for non-unit-test checks: ${VERSION}"
+          echo "::set-output name=version::${VERSION}"
+
+  php-setup:
+    name: Customize PHP setup
+    runs-on: ubuntu-latest
+    needs: [config]
+    outputs:
+      extensions: ${{ steps.php-setup.outputs.extensions }}
+      ini: ${{ steps.php-setup.outputs.ini }}
+    env:
+      CONFIG_JSON: ${{ needs.config.outputs.json }}
+    steps:
+      - id: php-setup
+        run: |
+          JQ_SCRIPT_TEMPLATE='def casttoarray(x): if x|type=="array" then x else [] end ; casttoarray(.php.extensions) + DEFAULTS'
+          JQ_SCRIPT=${JQ_SCRIPT_TEMPLATE/DEFAULTS/$PHP_DEFAULT_EXTENSIONS}
+          EXTENSIONS=$(echo "${CONFIG_JSON}" | jq -c "$JQ_SCRIPT")
+          echo "Using PHP extensions: ${EXTENSIONS}"
+          echo "::set-output name=extensions::${EXTENSIONS}"
+          JQ_SCRIPT_TEMPLATE='def casttoarray(x): if x|type=="array" then x else [] end ; casttoarray(.php.ini) + DEFAULTS'
+          JQ_SCRIPT=${JQ_SCRIPT_TEMPLATE/DEFAULTS/$PHP_DEFAULT_INI}
+          INI=$(echo "${CONFIG_JSON}" | jq -c "$JQ_SCRIPT")
+          echo "Using PHP INI settings: ${INI}"
+          echo "::set-output name=ini::${INI}"
+
+  unit-tests:
+    name: Unit Tests
+    runs-on: ubuntu-latest
+    needs: [config, php-setup, supported-versions-matrix, dependency-versions-matrix]
+
+    strategy:
+      fail-fast: false
+      matrix:
+        php-version: ${{ fromJson(needs.supported-versions-matrix.outputs.version) }}
+        deps-version: ${{ fromJson(needs.dependency-versions-matrix.outputs.version) }}
+        include:
+          - php-version: "8.0"
+            composer-args: --ansi --no-interaction --no-progress --prefer-dist --ignore-platform-req=php
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Setup PHP
+        uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php-version }}
+          extensions: ${{ join(fromJSON(needs.php-setup.outputs.extensions), ', ') }}
```

Comment:
Many of the `join(fromJSON(`... operations are because the previous actions did not pre-bake some values for us: perhaps if we had an action that implicitly extracted env vars (see https://github.com/marketplace/actions/github-environment-variables-action for inspiration) we could simplify a lot of this